### ISSUE TITLE
Unable to parse files with UTF-16 data

### ISSUE THREAD

**Issue by aswiniy** [antchfx/xmlquery#105]
Hi,
I am trying to parse a file that contains UTF-16 data. 
ex XML:
<?xml version="1.0" encoding="utf-16"?>
<Manifest baseLocation="downloads.dell.com" baseLocationAccessProtocols="HTTPS" dateTime="2023-04-24T01:09:20+05:30" identifier="ff3fea43-b011-4de8-924f-17c62f3daeb1" releaseID="D0JP9" version="23.04.28" predecessorID="66f24aa4-7d67-483e-884d-c8ec12220f20">
....
....
</Manifest>

When I try to parse with XMLquery, I am unable to read any elements. the contents of doc contains only the header element and the rest of the content as Chinese characters.

Below is my cod, I tried to read in 3 different ways .. all have the same problem... 
please help me if I need any corrections in my code.

func main() {
	filePath := "C:\\Users\\Aswini_Yallanki\\Downloads\\catalog_files\\Catalog\\Catalog1.xml"
	xmlData, err := os.ReadFile(filePath)
	if err != nil {
		fmt.Println("Error reading XML file:", err.Error())
		return
	}
	
//1. pass data as is
	doc, err := xmlquery.Parse(bytes.NewReader(xmlData))
	if err != nil {
		fmt.Println("Error parsing XML:", err)
		return
	}
	
//2.convert to utf8 and pass
	decoder := unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM).NewDecoder()
	utf8Data, err := decoder.Bytes(xmlData)
	if err != nil {
		fmt.Println("Error converting UTF-16 to UTF-8:", err)
		return
	}
// 3.convert to UTF-8 using charset reader
	str := string(xmlData)
	fmt.Println(str)
	reader := strings.NewReader(str)
	newReader, err := charset.NewReader(reader, "utf-16")
	doc, err := xmlquery.Parse(newReader)

}

**Comment by zhengchun** [comment]
What is you xpath express? Try call doc.OutXML() method to output all xml
for seeing whether pase is good.

aswiniy ***@***.***> 于 2023年6月9日周五 上午10:16写道：

**Comment by zhengchun** [comment]
Sorry，you mean not parse xml with utf16 encode. Can you put xml content to
go file and trying parse? Not read from xml file.

ZHENG CHUN ***@***.***> 于 2023年6月9日周五 下午2:43写道：

**Comment by aswiniy** [comment]
Hi @zhengchun Thank you for responding. Here I am attaching the file. 
[Catalog (4).gz](https://github.com/antchfx/xmlquery/files/11707966/Catalog.4.gz)

I have extracted this zip file to an XML file to use in my program.

When I print the string(xmlData) , I get it logged properly in console.
After calling xmlquery.Parse(newReader) , i couldn't read read any tags from the doc. 

xmlquery.QueryAll(doc, "//Display")
xmlquery.FindOne(doc, "//SoftwareBundle")

All these gives nil response , with out any error.

When I did debug the doc element,lastelement  reference has kind of Chinese character data in it. Screen shot attached for your reference. 

![Screenshot 2023-06-09 115247](https://github.com/antchfx/xmlquery/assets/6285442/1e66851d-c881-4f9a-ab85-c94285efe8dc)

**Comment by aswiniy** [comment]
When i try to print doc.OutputXML , This is the response i got.

![Screenshot 2023-06-09 120114](https://github.com/antchfx/xmlquery/assets/6285442/cfc9e989-3afb-4a46-93e5-665981c71dc1)

**Comment by zhengchun** [comment]
https://github.com/antchfx/xmlquery/commit/ce135a7fa3b0c3182dcc074c584a43a612ac2451

Test code:

```go
f, _ := os.Open(`Catalog.xml`)
defer f.Close()

// Create a UTF-16 to UTF-8 transformer
utf16ToUtf8Transformer := unicode.UTF16(unicode.LittleEndian, unicode.IgnoreBOM).NewDecoder()

// Wrap the UTF-16 Reader with the transformer to get a UTF-8 Reader
utf8Reader := transform.NewReader(f, utf16ToUtf8Transformer)

// Sets `CharsetReader` defined how to transform UTF-8 character
options := xmlquery.ParserOptions{
Decoder: &xmlquery.DecoderOptions{
	CharsetReader: func(charset string, input io.Reader) (io.Reader, error) {
		return input, nil
	},
},
}
doc, err := xmlquery.ParseWithOptions(utf8Reader, options)
if err != nil {
panic(err)
}
channel := xmlquery.FindOne(doc, "//SoftwareBundle")
if channel != nil {
if n := channel.SelectElement("Name"); n != nil {
	fmt.Printf("title: %s\n", n.InnerText())
}
} else {
fmt.Println("Element Not found")
}
```

Test success!

**Comment by aswiniy** [comment]
Hi @zhengchun 

Thank you very much for the quick fix. It works. :)

**Comment by Aemot** [comment]
I know this ticket is closed but I do not know of anywhere else to ask this question.  I pulled the latest tagged version (1.3.16) and am not seeing these changes.  When might these changes be added to an official release?  Thank you.

**Comment by zhengchun** [comment]
@Aemot , hello, the version of v1.3.16 not fix this bug. I released a new version v1.3.17 to supports UTF-16 XML.

https://github.com/antchfx/xmlquery/releases/tag/v1.3.17
